receiver/googlecloudspannerreceiver/internal/filterfactory/filterbymetricbuilder.go

```diff
@@ -0,0 +1,154 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filterfactory
+
+import (
+	"errors"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudspannerreceiver/internal/filter"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudspannerreceiver/internal/metadata"
+)
+
+type filterByMetricBuilder struct {
+	logger *zap.Logger
+	config *ItemFilterFactoryConfig
+}
+
+func (b filterByMetricBuilder) buildFilterByMetricZeroTotalLimit() map[string]filter.ItemFilter {
+	filterByMetric := make(map[string]filter.ItemFilter)
+	nopItemFilter := filter.NewNopItemCardinalityFilter()
+
+	for _, metadataItem := range b.config.MetadataItems {
+		for _, metricValueMetadata := range metadataItem.QueryMetricValuesMetadata {
+			metricFullName := metadataItem.MetricNamePrefix + metricValueMetadata.Name()
+			filterByMetric[metricFullName] = nopItemFilter
+		}
+	}
+
+	return filterByMetric
+}
+
+func (b filterByMetricBuilder) buildFilterByMetricPositiveTotalLimit() (map[string]filter.ItemFilter, error) {
+	filterByMetric := make(map[string]filter.ItemFilter)
+	groupedItems := groupByCardinality(b.config.MetadataItems)
+
+	// Handle metric groups with low cardinality
+	lowCardinalityGroups := groupedItems[false]
+	newTotalLimit, err := b.handleLowCardinalityGroups(lowCardinalityGroups, b.config.TotalLimit, filterByMetric)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle metric groups with high cardinality
+	highCardinalityGroups := groupedItems[true]
+	newTotalLimit, err = b.handleHighCardinalityGroups(highCardinalityGroups, newTotalLimit, filterByMetric)
+	if err != nil {
+		return nil, err
+	}
+
+	b.logger.Debug("Remaining total limit after cardinality limits calculation",
+		zap.Int("remainingTotalLimit", newTotalLimit))
+
+	return filterByMetric, nil
+}
+
+func (b filterByMetricBuilder) handleLowCardinalityGroups(groups []*metadata.MetricsMetadata, remainingTotalLimit int,
+	filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	if len(groups) == 0 {
+		return remainingTotalLimit, nil
+	}
+
+	limitPerMetricByTimestamp := b.config.ProjectAmount * b.config.InstanceAmount * b.config.DatabaseAmount
+	// For low cardinality metrics total limit is equal to limit by timestamp
+	totalLimitPerMetric := limitPerMetricByTimestamp
```

Comment:
The reassignment shouldn't be needed

Reply:
I agree, added this for better readability - will remove.

Reply:
Done.